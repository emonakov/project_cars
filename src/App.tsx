import React, { lazy, Suspense } from 'react';
import { ThemeProvider } from 'styled-components';
import { BrowserRouter as Router, Route, Switch } from 'react-router-dom';

import ContextProvider from './components/Context';
import Header from './components/Header';
import Footer from './components/Footer';
import Fallback from './components/Fallback';
import theme from './config/theme';

const Home = lazy(() => import('./components/Page/Home'));
const CarDetails = lazy(() => import('./components/Page/CarDetails'));

const App: React.FC = () => (
    <ThemeProvider theme={theme}>
        <Router>
            <Suspense fallback={<Fallback />}>
                <Header />
                <ContextProvider>
                    <Switch>
                        <Route path="/" exact component={Home} />
                        <Route path="/car/:stockId(\d+)" exact component={CarDetails} />
                    </Switch>
                </ContextProvider>
                <Footer />
            </Suspense>
        </Router>
    </ThemeProvider>
);

export default App;
