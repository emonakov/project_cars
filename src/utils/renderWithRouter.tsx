import React from 'react';
import { Router } from 'react-router-dom';
import { render } from '@testing-library/react';
import { createMemoryHistory } from 'history';

const history = createMemoryHistory();

const renderWithRouter = (Component: JSX.Element, initialRoute = '/') => {
    history.push(initialRoute);

    return render(
        <Router history={history}>
            {Component}
        </Router>,
    );
};

export default renderWithRouter;
