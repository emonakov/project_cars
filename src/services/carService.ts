import axios from 'redaxios';

import { buildApiUrl } from '../utils/api';

export const fetchCar = async (stockId: string): Promise<any> => {
    const { data } = await axios.get(buildApiUrl(`api/cars/${stockId}`));
    const { car } = JSON.parse(data);

    return car;
};
