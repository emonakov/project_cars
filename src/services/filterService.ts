import axios from 'redaxios';

import { buildApiUrl } from '../utils/api';

export const fetchFilters = async (): Promise<any> => {
    const manufacturersUrl = buildApiUrl('api/manufacturers');
    const colorsUrl = buildApiUrl('api/colors');

    const [manResponse, colResponse] = await Promise.all([axios.get(manufacturersUrl), axios.get(colorsUrl)]);
    const [{ manufacturers }, { colors }] = [JSON.parse(manResponse.data), JSON.parse(colResponse.data)];

    return { manufacturers, colors };
};

export const fetchCarsWithFilters = async (options: any[]): Promise<any> => {
    const carsUrl = options
        .filter(([, paramValue]) => Boolean(paramValue))
        .reduce((acc, [paramKey, paramValue]) => `${acc}&${paramKey}=${paramValue}`, 'api/cars?');

    const { data } = await axios.get(buildApiUrl(carsUrl));

    return JSON.parse(data);
};
