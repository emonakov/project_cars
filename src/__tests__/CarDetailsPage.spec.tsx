import React from 'react';
import { wait, fireEvent } from '@testing-library/react';

import App from '../App';
import { fetchCarsWithFilters } from '../services/filterService';
import { fetchCar } from '../services/carService';
import renderWithRouter from '../utils/renderWithRouter';
import cars from './__mock__/cars.mock.json';

jest.mock('../services/filterService');
jest.mock('../services/carService');

describe('Tests CarDetails page', () => {
    beforeEach(() => {
        (fetchCarsWithFilters as jest.Mock).mockImplementation((callback: Function) => {
            callback(cars.Audi);
        });

        (fetchCar as jest.Mock).mockImplementation((callback: Function) => {
            callback(cars.Audi.cars[0]);
        });
    });

    afterEach(() => {
        jest.clearAllMocks();
        window.localStorage.clear();
    });

    it('Renders the initial page and navigates to the car view page', async () => {
        const { container, getByText, getByTestId } = renderWithRouter(<App />);
        await wait();

        const carLink = getByTestId('car-80765');
        fireEvent.click(carLink);
        await wait();

        expect(getByText(/80765/)).toBeInTheDocument();
        expect(getByTestId('fav-button')).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });

    it('Should save the car in the local storage', () => {
        const { getByTestId } = renderWithRouter(<App />, '/car/80765');
        const favButton = getByTestId('fav-button');

        fireEvent.click(favButton);
        expect(favButton).toHaveTextContent(/remove/i);
        expect(localStorage.getItem('favCars')).not.toBeNull();
        expect(localStorage.getItem('favCars')).not.toBe('[]');
    });

    it('Should remove the car from the local storage', () => {
        const { getByTestId } = renderWithRouter(<App />, '/car/80765');
        const favButton = getByTestId('fav-button');

        fireEvent.click(favButton);
        fireEvent.click(favButton);
        expect(localStorage.getItem('favCars')).toBe('[]');
    });

    it('Should show remove button if the car is in the local storage', () => {
        window.localStorage.setItem('favCars', JSON.stringify([cars.Audi.cars[0]]));
        const { getByTestId } = renderWithRouter(<App />, '/car/80765');
        const favButton = getByTestId('fav-button');

        expect(favButton).toHaveTextContent(/remove/i);
    });

    it('Throws the error if something happened in the service ', async () => {
        (fetchCar as jest.Mock).mockImplementation((callback: Function) => {
            callback(null, new Error('something went wrong'));
        });

        const { getByText } = renderWithRouter(<App />, '/car/80765');
        await wait();

        expect(getByText(/something went wrong/i)).toBeInTheDocument();
    });
});
