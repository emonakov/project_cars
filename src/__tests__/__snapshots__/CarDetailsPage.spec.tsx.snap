// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests CarDetails page Renders the initial page and navigates to the car view page 1`] = `
<div>
  <header
    style=""
  >
    <div
      class="MuiContainer-root sc-AxjAm sc-AxheI kTVrpe MuiContainer-maxWidthLg"
    >
      <div
        class="MuiGrid-root sc-AxiKw bBNIh MuiGrid-container MuiGrid-spacing-xs-3 MuiGrid-align-items-xs-center MuiGrid-justify-xs-space-between"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-md-4"
        >
          <h1
            class="MuiTypography-root sc-AxhUy gpIutg MuiTypography-h1"
          >
            <a
              aria-current="page"
              class="sc-AxirZ iLjBBz sc-AxhCb sc-AxgMl kETrBh active"
              href="/"
            >
              Car Sales
            </a>
          </h1>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-md-4"
        >
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 MuiGrid-justify-xs-flex-end"
          >
            <a
              class="sc-AxirZ iLjBBz sc-AxhCb clXXet"
              href="/car/fav"
            >
              Favorites
            </a>
            <a
              aria-current="page"
              class="sc-AxirZ iLjBBz sc-AxhCb clXXet active"
              href="/"
            >
              My orders
            </a>
            <a
              aria-current="page"
              class="sc-AxirZ iLjBBz sc-AxhCb clXXet active"
              href="/"
            >
              Sell
            </a>
          </div>
        </div>
      </div>
    </div>
  </header>
  <div
    class="MuiContainer-root sc-AxjAm sc-fznWqX eipsae MuiContainer-maxWidthLg"
  >
    <section
      class="sc-fzoiQi kKCChg"
    />
  </div>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg"
  >
    <section
      class="sc-fzoXzr bYAKAE"
    >
      <div
        class="MuiCardContent-root"
      >
        <h4
          class="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom"
        >
          R8
        </h4>
        <p
          class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom MuiTypography-displayBlock"
        >
          Stock # 80765
          <span
            class="sc-fzplWN beffhE"
          >
             - 176,817 km
          </span>
          <span
            class="sc-fznyAO looylO"
          >
             - Petrol - white
          </span>
        </p>
        <p
          class="MuiTypography-root MuiTypography-body2"
        >
          This car is currently available and can be delivered as soon as tomorrow morning. Please be aware that delivery times shown in this page are not definitive and may change due to bad weather conditions
        </p>
      </div>
      <div
        class="MuiCardContent-root sc-fzpjYC dWnwVz"
      >
        <section
          class="sc-fznxsB iybvfE"
        >
          <span
            class="MuiTypography-root MuiTypography-caption"
          >
            If you like this car, click the button and save it in your collection of favorite items.
          </span>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained sc-fzoLsD jtNKkn MuiButton-containedPrimary"
            data-testid="fav-button"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Save
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </section>
      </div>
    </section>
  </div>
  <footer
    class="sc-Axmtr jAtLrE"
    style=""
  >
    ©    ¯\\_(ツ)_/¯    2020
  </footer>
</div>
`;
